A Docker-based interactive Jupyterlab powered by GPU for artificial intelligence in Galaxy

anuprulez opened the pull request:
This PR adds a training material for image segmentation prediction using COVID CT scans in a Jupyterlab notebook. The Jupyterlab notebook is integrated into Galaxy Europe as an interactive tool. The standard Jupyterlab has been augmented with many features such as Git integration, Elyra AI, GPU support and runs inside a [Docker container](https://github.com/anuprulez/ml-jupyter-notebook). Moreover, remote processing of long-running deep learning training tasks is also possible via running a Galaxy tool that uploads scripts and datasets to Galaxy history and executes scripts on a remote cluster. The entire analysis comes from the paper: https://www.sciencedirect.com/science/article/pii/S2666990021000069 

Ground truth and predicted COVID CT scan masks:
![true_pred_masks](https://user-images.githubusercontent.com/3022518/159055875-1ef5b1d0-d515-4c5d-9e5b-01e57d108fc0.png)

Remote processing:
![finished_history_remote_ai](https://user-images.githubusercontent.com/3022518/159055791-32fadcf9-1eb8-49cb-af52-777bf4dcac57.png)

3D protein structure (for a shorter SARS-CoV-2 spike protein sequence of length 300)

![3D_300_L_protein](https://user-images.githubusercontent.com/3022518/160106764-7c402d03-014b-45f8-b210-4676b7893983.png)

A bit of work is still left for this tutorial such as the addition of another use-case of integration with AlphaFold2. Would love to get some comments from @kxk302 and others as well. Thanks!!

Edit on 25.03.22: It is ready to be reviewed.

ping @bgruening

anuprulez commented:
This PR is ready to be reviewed. @kxk302 would you have time to review this? Thanks!

kxk302 commented:
Hi @anuprulez 
Yes, will do.

kxk302 commented:
Sorry @anuprulez, did not get to review this today. Will review on Monday!
Have a great weekend.

anuprulez commented:
@kxk302 no worries, have a nice weekend

hexylena commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,181 @@
+layout: tutorial_hands_on
+
+title: A Docker-based interactive Jupyterlab powered by GPU for artificial intelligence in Galaxy
+zenodo_link: https://zenodo.org/record/6091361#.Ygu4gIzMI5k
```
```suggestion
zenodo_link: https://zenodo.org/record/6091361
```

hexylena commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,181 @@
+
+---
+
+{:.no_toc}
```
```suggestion
```

hexylena commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,181 @@
+
+
+## Jupyterlab features
+[Jupyterlab](https://jupyterlab.readthedocs.io/en/stable/) notebook has been augmented with several useful features and together they make it ready-to-use for quick prototyping of AI projects. Being able to **serve online** makes it convenient to share it with other researchers and users. GPUs have accelerated AI research, especially deep learning. Therefore, the backend of the Jupyterlab is **po ...
```
`[<<open notebook>>/elyra/METABRIC_ML.pipeline]`

looks odd here

hexylena commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,181 @@
+
+![Unet architecture](../../images/Unet.jpg "Architecture of Unet neural network for image segmentation. The name "Unet" resembles the shape of its architecture as 'U'")
+
+In this tutorial, we will use the dataset of CTs scans and their respective masks to train an Unet neural network model. The model learns to map the infected regions in the CT scans to their masks. For prediction, the trained model is given unseen CT scans and for each CT scan, it predicts infected regions or masks. For this experiment, we will use Jupyterlab for cloning the notebooks from Github ...
```
```suggestion
In this tutorial, we will use the dataset of CTs scans and their respective masks to train an Unet neural network model. The model learns to map the infected regions in the CT scans to their masks. For prediction, the trained model is given unseen CT scans and for each CT scan, it predicts infected regions or masks. For this experiment, we will use Jupyterlab for cloning the notebooks from Github that contain scripts for downloading data, creating and training Unet model and prediction tasks. Data (CT scans and also the trained model) required for the notebooks can be downloaded from [Zenodo](https://zenodo.org/record/6091361). The model can either be trained in the Jupyterlab or can be sent to Galaxy's cluster for remote processing. After remote processing, the created datasets such as the trained model become available in the new Galaxy history.
```

hexylena commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,181 @@
+In this tutorial, we have discussed several features of the Jupyterlab notebook in Galaxy and used a few of them to perform an image segmentation task on COVID CT scans by training Unet, a deep learning model. In addition, we have used it for predicting 3D structure of a protein sequences which is a memory-intensive and time-consuming task. The ready-to-use infrastructure that we used here for de ...
+
+
+{:.no_toc}
```
```suggestion
```

hexylena reviewed (commented):
overall formatting and image captions look ok! Thanks @anuprulez

kxk302 commented:
Hi @anuprulez 

I am trying to view your tutorial via gitpod (https://training.galaxyproject.org/training-material/topics/contributing/tutorials/gitpod/tutorial.html). If you look at instructions under 'Already have a fork of the GTN?'
you will see that your fork must have a main branch (not master) and it must be up-to-date. Could you please rename your master branch to main and make sure its up-to-date, so I can use gitpod to view your tutorial? Thanks!

hexylena commented:
@kxk302 if you check the very bottom left of gitpod you can change branches there to the one this PR is coming from.

![image](https://user-images.githubusercontent.com/458683/160420118-fcebe829-8dca-445b-b848-706143d6f222.png)

anuprulez commented:
@kxk302 I have updated the "master" branch in my fork to "main" (https://github.com/anuprulez/training-material) from which this current feature branch was created and this feature branch is also up-to-date with the `galaxyproject/training-material:main.`
I guess this should work :+1:

kxk302 commented:
> @kxk302 I have updated the "master" branch in my fork to "main" (https://github.com/anuprulez/training-material) from which this current feature branch was created and this feature branch is also up-to-date with the `galaxyproject/training-material:main.` I guess this should work 👍

Thanks @anuprulez !

anuprulez commented:
@kxk302 was it possible for you to install the branch? thanks but I am not able to see your review comments here

kxk302 commented:
> @kxk302 was it possible for you to install the branch? thanks but I am not able to see your review comments here

@anuprulez if you go tutorial.md in Files Changed tab, you should see my embedded comments. Let me know if you have any issues

anuprulez commented:
@kxk302 I am not able to see your comments. Here is the link to the file that I am looking at: https://github.com/galaxyproject/training-material/pull/3278/files?file-filters%5B%5D=.md&show-viewed-files=true&w=1

![md_file](https://user-images.githubusercontent.com/3022518/160615920-5d19954f-a7fa-4e5c-8e85-1d8096ac3a99.png)

Also, `Show comments` filter is checked for this file but I don't see any comment. Do I miss something? Can you provide the line number of one of your comments? Thanks

anuprulez commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,176 @@
+In this tutorial, we will use the dataset of CTs scans and their respective masks to train an Unet neural network model. The model learns to map the infected regions in the CT scans to their masks. For prediction, the trained model is given unseen CT scans and for each CT scan, it predicts infected regions or masks. For this experiment, we will use Jupyterlab for cloning the notebooks from Github ...
+
+## Protein 3D structure
+Understanding the structure of proteins provides insights into their functions. But, only a few proteins have known structures out of billions of known proteins. To advance into the direction of predicting their 3D structures only from their sequences, [AlphaFold2](https://www.nature.com/articles/s41586-021-03819-2) has made a breakthrough to predict their structures with high accuracy. However,  ...
```
I have moved the references to the bottom of the tutorial. Thanks!

anuprulez commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,176 @@
+
+
+## CT scans and masks
+Figure 4 shows CT scans of infected lungs (top row) and borders around the infected regions have been drawn with red color (middle row). In the last row respective masks, from the CT scans, have been taken out denoting the infected regions. These masks are the "segmented" regions from the corresponding CT scans. While creating the dataset for training the deep learning model (Unet), the CT scans  ...
```
Unet has a small section in the tutorial. Thanks!

anuprulez commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,176 @@
+
+
+## COVID CT scan
+Segmenting lung CT scans task to locate infected regions has been actively proposed to augment the RT-PCR testing for initial screening of COVID infection in humans. Deep learning has been used to predict these regions with high accuracy in studies such as [Inf-net: Automatic COVID-19 lung infection segmentation from CT images](https://ieeexplore.ieee.org/document/9098956), [JCS: An explainable C ...
```
Moved the references to the bottom.

anuprulez commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,176 @@
+
+![Elyra AI pipeline](../../images/elyra_ai_pipeline.png "An example ML pipeline created using Elyra AI in a notebook. Each connected block of the pipeline is a different notebook")
+
+There are several additional packages suited for performing machine learning tasks such as [Open-CV](https://github.com/opencv/opencv) and [Scikit-Image](https://scikit-image.org/) for image processing, [NiBabel](https://nipy.org/nibabel/) package for processing images files. These packages are available to researchers and users in this notebook. Many other features such as GPU utilization dashbo ...
```
done!

anuprulez commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,176 @@
+Understanding the structure of proteins provides insights into their functions. But, only a few proteins have known structures out of billions of known proteins. To advance into the direction of predicting their 3D structures only from their sequences, [AlphaFold2](https://www.nature.com/articles/s41586-021-03819-2) has made a breakthrough to predict their structures with high accuracy. However,  ...
+
+# Reproduce results from recent publications using Jupyterlab infrastructure in Galaxy
+In this tutorial, we will a few features of Jupyterlab to create and train a Unet deep learning model and then, predict segmented regions in COVID CT scans using the Unet trained model. In addition, we will predict 3D structure of protein sequence using ColabFold, a faster implementation of AlphaFold2.
```
updated the word, thanks!

anuprulez commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,176 @@
+> 
+> ![Galaxy history](../../images/finished_history_remote_ai.png "Galaxy history showing finished datasets after remote training on a Galaxy cluster")
+> 
+> **Note**: the training may take longer depending on how busy Galaxy's queueing is as it sends the training task to be done on a Galaxy cluster. Therefore, this feature should be used when the training task is expected to run for at least a couple of hours. The training time is higher because a large Docker container is downloaded on the assigned cluster and only then, the training task can proc ...
```
Done!

anuprulez commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,176 @@
+
+
+# Conclusion
+In this tutorial, we have discussed several features of the Jupyterlab notebook in Galaxy and used a few of them to perform an image segmentation task on COVID CT scans by training Unet, a deep learning model. In addition, we have used it for predicting 3D structure of a protein sequences which is a memory-intensive and time-consuming task. The ready-to-use infrastructure that we used here for de ...
```
yes, now it has two paragraphs, thanks!

anuprulez commented:
@kxk302 thanks for reviewing the tutorial. I have added fixes for the review comments. I have added smaller sections for those that were bigger and organised them as more coherent sections. Would be happy to know your further suggestions.

kxk302 commented on topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:
```diff
@@ -0,0 +1,176 @@
+
+> ### {% icon hands_on %} Hands-on: Run notebooks in Jupyterlab
+>
+> 1. Download and save datasets in the notebook using **"1_fetch_datasets.ipynb"** notebook. It will also create all the necessary folders. It downloads two datasets, one "h5" file containing many matrices as sub-datasets belonging to training data, training labels, validation data, validation labels, test data and test labels. These sub-datasets are stored in different variables after reading th ...
```
Thanks Anup, they look very good

kxk302 commented:
@anuprulez the changes look great. Thank you!

anuprulez commented:
can this be merged once the build passes? Thanks!

bgruening commented:
I'm halfway through ...

anuprulez commented:
> I'm halfway through ...

Please let me know if you have any further comments to be fixed/added :)

bgruening reviewed (approved):
This is a great first version @anuprulez 
Thanks a lot!

anuprulez commented:
Thanks to all @bgruening @hexylena @kxk302
